### PR TITLE
Added Drag & Drop Reordering to Shopping List Card.

### PR DESCRIPTION
**Pull request by ShaneQi**
## Proposed change

Add drag & drop reordering to shopping list card.

The Core PR: https://github.com/home-assistant/core/pull/41585

Kapture 2020-10-10 at 01 25 26

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

N/A

## Additional information

- This PR fixes or closes issue: N/A
- This PR is related to issue: N/A
- Link to documentation pull request: N/A

## Checklist

- [X] The code change is tested and works locally.
- [X] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [N/A] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/panels/lovelace/cards/hui-shopping-list-card.ts [hunk 1]
```diff
@@ -120,28 +131,27 @@ class HuiShoppingListCard extends SubscribeMixin(LitElement)
...
           ></paper-input>
+          <ha-icon
+            class="reorderButton"
+            icon="hass:sort"
+            .title=${this.hass!.localize(
+              "ui.panel.lovelace.cards.shopping-list.reorder_items"
+            )}
+            @click=${this._toggleReorder}
+          >
+          </ha-icon>
```
Comment (zsarnett) [inline src/panels/lovelace/cards/hui-shopping-list-card.ts:142]: ha-svg-icon

File: src/panels/lovelace/cards/hui-shopping-list-card.ts [hunk 2]
```diff
@@ -189,6 +199,44 @@ class HuiShoppingListCard extends SubscribeMixin(LitElement)
...
   }
 
+  private _renderItems(items: ShoppingListItem[]) {
+    return html`
+      ${repeat(
+        items,
+        (item) => item.id,
+        (item) =>
+          html`
+            <div class="editRow" item-id=${item.id}>
```
Comment (zsarnett) [inline src/panels/lovelace/cards/hui-shopping-list-card.ts:209]:
```suggestion
            <div class="editRow" .itemId=${item.id}>
```

Comment (zsarnett) [reply src/panels/lovelace/cards/hui-shopping-list-card.ts:209]: actually do you need this

File: src/panels/lovelace/cards/hui-shopping-list-card.ts [hunk 3]
```diff
@@ -250,6 +298,50 @@ class HuiShoppingListCard extends SubscribeMixin(LitElement)
     }
   }
 
+  private async _toggleReorder() {
+    if (!Sortable) {
+      const [sortableImport] = await Promise.all([
+        import("sortablejs/modular/sortable.core.esm"),
+      ]);
```
Comment (zsarnett) [inline src/panels/lovelace/cards/hui-shopping-list-card.ts:305]: No need for the promise all

File: src/panels/lovelace/cards/hui-shopping-list-card.ts (hunk 2 above)
Comment (ShaneQi) [reply src/panels/lovelace/cards/hui-shopping-list-card.ts:209]: > actually do you need this

@zsarnett  Yes, [`Sortable` uses it](https://github.com/home-assistant/frontend/pull/7296/files#diff-51b956cdb853a2c20ff93c5cd7d3963fR322), so that we can conveniently [get reordered item ids array](https://github.com/home-assistant/frontend/pull/7296/files#diff-51b956cdb853a2c20ff93c5cd7d3963fR325).

Comment (ShaneQi) [reply src/panels/lovelace/cards/hui-shopping-list-card.ts:209]: Well we actually could avoid this by mapping reordered `this._uncheckedItems` to item ids after [this](https://github.com/home-assistant/frontend/pull/7296/files#diff-51b956cdb853a2c20ff93c5cd7d3963fR333).

Which way do you think is better?

File: src/panels/lovelace/cards/hui-shopping-list-card.ts (hunk 1 above)
Comment (ShaneQi) [reply src/panels/lovelace/cards/hui-shopping-list-card.ts:142]: I see that the whole file is still using `ha-icon`, I will migrate them all in this PR.

File: src/panels/lovelace/cards/hui-shopping-list-card.ts (hunk 3 above)
Comment (ShaneQi) [reply src/panels/lovelace/cards/hui-shopping-list-card.ts]: fixed

File: src/panels/lovelace/cards/hui-shopping-list-card.ts (hunk 2 above)
Comment (zsarnett) [reply src/panels/lovelace/cards/hui-shopping-list-card.ts:209]: What you are doing is fine. I forgot about that for sortable.

File: src/panels/lovelace/cards/hui-shopping-list-card.ts [hunk 4]
```diff
@@ -250,6 +298,48 @@ class HuiShoppingListCard extends SubscribeMixin(LitElement)
...
+    if (this._reordering) {
+      this._createSortable();
+    } else {
+      this._sortable?.destroy();
+      this._sortable = undefined;
+    }
+  }
+
+  private _createSortable() {
+    this._sortable = new Sortable(this.shadowRoot!.getElementById("sortable"), {
```
Comment (zsarnett) [inline src/panels/lovelace/cards/hui-shopping-list-card.ts:317]: We should use a Query Decorator here and cache it.

So this would be a new class variable
```ts
@query("#sortable") private _sortableEl?: HTMLElement;
```

**Comment by zsarnett** [comment]
Also, you may need to rebase your PR again. Not sure about the build error you are getting. Something got messed up with the localize

**Comment by ShaneQi** [comment]
> Also, you may need to rebase your PR again. Not sure about the build error you are getting. Something got messed up with the localize

I've discussed the build error with **balloob** in #devs_frontend last night. Some localization issues on Lokalise broke `dev`, therefore CI isn't passing here (CI merges PR into dev locally before running checks).

I fixed the localization issues on Lokalise, we should be good once it's pulled into repo today. After that I will push another commit to address your comment(s) which will re-trigger CI checks. I'm sure it will pass, otherwise I will rebase.

File: src/panels/lovelace/cards/hui-shopping-list-card.ts [hunk 5]
```diff
@@ -331,9 +336,11 @@ class HuiShoppingListCard extends SubscribeMixin(LitElement)
         );
         this._renderEmptySortable = true;
         await this.updateComplete;
-        const container = this.shadowRoot!.getElementById("sortable")!;
-        while (container.lastElementChild) {
-          container.removeChild(container.lastElementChild);
+        const container = this._sortableEl;
```
Comment (zsarnett) [inline src/panels/lovelace/cards/hui-shopping-list-card.ts:339]: why create this?

File: src/panels/lovelace/cards/hui-shopping-list-card.ts [hunk 6]
```diff
@@ -331,9 +336,11 @@ class HuiShoppingListCard extends SubscribeMixin(LitElement)
...
+        if (container) {
+          while (container.lastElementChild) {
+            container.removeChild(container.lastElementChild);
+          }
         }
```
Comment (zsarnett) [inline src/panels/lovelace/cards/hui-shopping-list-card.ts]:
```suggestion
        while (this._sortableEl?.lastElementChild) {
          this._sortableEl.removeChild(this._sortableEl.lastElementChild);
        }
```

File: src/panels/lovelace/cards/hui-shopping-list-card.ts (hunk 5 above)
Comment (ShaneQi) [reply src/panels/lovelace/cards/hui-shopping-list-card.ts]: > why create this?

Because this element will be used multiple times below, I thought I should use a local variable to hold it, so that I won't be calling the `@query` getter too many times without knowing how expensive the `@query` getter is. It's more like an instinct, not really a solid reason.

But I change it to what you suggested, it's obviously cleaner. And since you suggested it, what I was thinking is probably unnecessary.

Comment (zsarnett) [reply src/panels/lovelace/cards/hui-shopping-list-card.ts]: Its also cached now which is very nice.

**Comment by balloob** [comment]
I'm a little hesitant to always show the drag and drop handle for each row. Maybe we can add a "reorder" menu item to turn on showing the handle, with a "finish" button ?

**Comment by zsarnett** [comment]
He has an icon that does this

**Comment by ShaneQi** [comment]
Like Zack mentioned, this button toggles the drag & drop handles.

image

File: src/panels/lovelace/cards/hui-shopping-list-card.ts [hunk 7]
```diff
@@ -23,11 +25,15 @@ import {
   fetchItems,
   ShoppingListItem,
   updateItem,
+  reorderItems,
 } from "../../../data/shopping-list";
 import { SubscribeMixin } from "../../../mixins/subscribe-mixin";
 import { HomeAssistant } from "../../../types";
 import { LovelaceCard, LovelaceCardEditor } from "../types";
 import { SensorCardConfig, ShoppingListCardConfig } from "./types";
+import { mdiDrag, mdiSort, mdiPlus, mdiNotificationClearAll } from "@mdi/js";
```
Comment (zsarnett) [inline src/panels/lovelace/cards/hui-shopping-list-card.ts:34]: Move to the top of the file under Guard import. Absolute imports should be above relative

Comment (ShaneQi) [reply src/panels/lovelace/cards/hui-shopping-list-card.ts]: Good to know 👍🏽. Fixed.

File: src/panels/lovelace/cards/hui-shopping-list-card.ts (hunk 4 above)
Comment (bramkragten) [reply src/panels/lovelace/cards/hui-shopping-list-card.ts:322]: We can't cache it as it is removed and readded when you switch reorder mode

File: src/panels/lovelace/cards/hui-shopping-list-card.ts [hunk 8]
```diff
@@ -250,6 +301,49 @@ class HuiShoppingListCard extends SubscribeMixin(LitElement)
...
+    if (this._reordering) {
+      this._createSortable();
+    } else {
+      this._sortable?.destroy();
+      this._sortable = undefined;
+    }
+  }
+
+  private _createSortable() {
+    this._sortable = new Sortable(this.shadowRoot!.getElementById("sortable"), {
```
Comment (bramkragten) [inline src/panels/lovelace/cards/hui-shopping-list-card.ts:322]:
```suggestion
    const sortableEl = this._sortableEl;
    this._sortable = new Sortable(sortableEl, {
```

File: src/panels/lovelace/cards/hui-shopping-list-card.ts [hunk 9]
```diff
@@ -250,6 +301,49 @@ class HuiShoppingListCard extends SubscribeMixin(LitElement)
...
+        // Move the shopping list item in memory.
+        this._uncheckedItems!.splice(
+          evt.newIndex,
+          0,
+          this._uncheckedItems!.splice(evt.oldIndex, 1)[0]
+        );
+        this._renderEmptySortable = true;
+        await this.updateComplete;
+        while (this._sortableEl?.lastElementChild) {
+          this._sortableEl.removeChild(this._sortableEl.lastElementChild);
```
Comment (bramkragten) [inline src/panels/lovelace/cards/hui-shopping-list-card.ts:340]:
```suggestion
        while (sortableEl.lastElementChild) {
          sortableEl.removeChild(sortableEl.lastElementChild);
```

File: src/panels/lovelace/cards/hui-shopping-list-card.ts [hunk 10]
```diff
@@ -51,6 +57,14 @@ class HuiShoppingListCard extends SubscribeMixin(LitElement)
 
   @internalProperty() private _checkedItems?: ShoppingListItem[];
 
+  @internalProperty() private _reordering = false;
+
+  @internalProperty() private _renderEmptySortable = false;
+
+  private _sortable?;
+
+  @query("#sortable", true) private _sortableEl?: HTMLElement;
```
Comment (bramkragten) [inline src/panels/lovelace/cards/hui-shopping-list-card.ts:66]:
```suggestion
  @query("#sortable") private _sortableEl?: HTMLElement;
```

File: src/panels/lovelace/cards/hui-shopping-list-card.ts [hunk 11]
```diff
@@ -250,6 +301,49 @@ class HuiShoppingListCard extends SubscribeMixin(LitElement)
...
+        reorderItems(this.hass!, this._sortable.toArray()).catch(() =>
+          this._fetchData()
+        );
+        // Move the shopping list item in memory.
+        this._uncheckedItems!.splice(
+          evt.newIndex,
+          0,
+          this._uncheckedItems!.splice(evt.oldIndex, 1)[0]
+        );
+        this._renderEmptySortable = true;
```
Comment (bramkragten) [inline src/panels/lovelace/cards/hui-shopping-list-card.ts:337]: I think we should also do a round of `_renderEmptySortable` when `this._uncheckedItems` is changed? What happenes when you add an item while you are sorting the list? And what happens when storing the new order (line 329) failed?

File: src/panels/lovelace/cards/hui-shopping-list-card.ts [hunk 12]
```diff
@@ -250,6 +301,49 @@ class HuiShoppingListCard extends SubscribeMixin(LitElement)
...
+
+  private _createSortable() {
+    this._sortable = new Sortable(this.shadowRoot!.getElementById("sortable"), {
+      animation: 150,
+      fallbackClass: "sortable-fallback",
+      dataIdAttr: "item-id",
+      handle: "ha-svg-icon",
+      onSort: async (evt) => {
+        reorderItems(this.hass!, this._sortable.toArray()).catch(() =>
+          this._fetchData()
```
Comment (bramkragten) [inline src/panels/lovelace/cards/hui-shopping-list-card.ts:329]: We should let the user know it failed and why we undo all the sorting work.

**Comment by ShaneQi** [comment]
Will work on addressing comments ASAP.

File: src/panels/lovelace/cards/hui-shopping-list-card.ts (hunk 4 above)
Comment (ShaneQi) [reply src/panels/lovelace/cards/hui-shopping-list-card.ts]: Good catch!

File: src/panels/lovelace/cards/hui-shopping-list-card.ts (hunk 11 above)
Comment (ShaneQi) [reply src/panels/lovelace/cards/hui-shopping-list-card.ts:338]: Tested the 2 scenarios without changing anything, both works fine.

When `this._uncheckedItems` is changed, the whole `sortable` div content will be re-rendered (at [line 150](https://github.com/home-assistant/frontend/pull/7296/files#diff-3f9c900506d8e374961fa5be5db01f33e7d214a7e3f02ecd9e0dcb825fc3b993R150)), so everything should be straight.

Comment (bramkragten) [reply src/panels/lovelace/cards/hui-shopping-list-card.ts:338]: That is not true, Lit uses comments to keep track of things, because with sorting we change the DOM but not the comments Lit loses track of some elements. That's why we clear the DOM and render empty and start over. That can cause duplicate items in the list when not rendering empty.

Comment (ShaneQi) [reply src/panels/lovelace/cards/hui-shopping-list-card.ts:338]: I'll do more digging tonight.

Comment (ShaneQi) [reply src/panels/lovelace/cards/hui-shopping-list-card.ts:338]: @bramkragten I took another look, I still think there is no need to explicitly 'do a round of `_renderEmptySortable` when this._uncheckedItems is changed'.

Because at [line 150](https://github.com/home-assistant/frontend/pull/7296/files#diff-3f9c900506d8e374961fa5be5db01f33e7d214a7e3f02ecd9e0dcb825fc3b993R150): `guard([this._uncheckedItems, this._renderEmptySortable]...` decides that when either one of the 2 values (`_uncheckedItems`, `_renderEmptySortable`) changes, the content of sortable `div` will be re-rendered ([doc](https://lit-html.polymer-project.org/guide/template-reference#guard)). 

When add an item while sorting the list, `_uncheckedItems` is changed therefore the content of sortable `div` will be re-rendered. Unless you are talking about the scenario that: the user adds an item after the user starts dragging and BEFORE the user puts it down, but I don't see how that is possible.
Also when storing the new order failed, `_uncheckedItems` is changed therefore the content of sortable `div` will be re-rendered.

In short: the list is messed up only after drag-n-drop, we do a round of `_renderEmptySortable` to re-render the list (also clean up the zombie elements). In any other cases, the change of `this._uncheckedItems` will always re-render list, no need to use `_renderEmptySortable`, also there is no zombie elements generated by drag-n-drop.

Comment (ShaneQi) [reply src/panels/lovelace/cards/hui-shopping-list-card.ts:342]: But while I was digging into it, I found another edge case:

1. Drag an item around and put the item back to its original position.
2. The Lit format is messed up by dragging around.
3. But the cleanup block (toggling `_renderEmptySortable)` won't be executed because putting the item back to its original position don't trigger `onSort` event.
4. Therefore the format won't be cleaned up and causing issues in other user interactions.

The bug:

Kapture 2020-11-27 at 16 00 25

The fix I put in is that instead of observing `onSort` event, we observe `onEnd` event, which will be triggered even if an item is dragged and put back to its original position. In the block, we check on `evt.oldIndex !== evt.newIndex` to decide if a websocket message needs to be sent.

After fix:

Kapture 2020-11-27 at 16 06 30

File: src/panels/lovelace/cards/hui-shopping-list-card.ts (hunk 12 above)
Comment (ShaneQi) [reply src/panels/lovelace/cards/hui-shopping-list-card.ts]: I will add a toast for the error message tonight.

File: src/panels/lovelace/cards/hui-shopping-list-card.ts (hunk 11 above)
Comment (bramkragten) [reply src/panels/lovelace/cards/hui-shopping-list-card.ts:342]: The problem is that re-rendering doesn't clear up the nodes that have no comment anymore.

Comment (ShaneQi) [reply src/panels/lovelace/cards/hui-shopping-list-card.ts:342]: @bramkragten now I see what your concern is, and it's a very valid point.

My thinking is: the no-comment-nodes are only generated after drag&drop, so we only need to clear them up after drag&drop. We are already doing that in the sortable's `onEnd` event. That should guarantee we always clear up the no-comment-nodes.

If you still don't think it's safe, I have no problem adding the clear-up when `this._uncheckedItems` changes. Let me know.

BTW I need to work on the toast to show errors, I forgot it last time but now I remember.

Comment (ShaneQi) [reply src/panels/lovelace/cards/hui-shopping-list-card.ts:342]: @bramkragten  Thoughts? Does above make sense to you? Or should I go ahead add the clear-up when `this._uncheckedItems` changes? I don't mind either way.

**Review by bramkragten — APPROVED** [review]

**Comment by ShaneQi** [comment]
Wanted to leave a comment that if there are issues in the future, please remind me, I'd be happy to help.
